### ISSUE TITLE
Learnable edge weight in GCNConv

### ISSUE THREAD

**Issue by hariharannatesh** [rusty1s/pytorch_geometric#2033]
## ❓ Questions & Help
Hi
Is it possible to set the edge_weight argument in GCNConv as a learnable parameter? I am attaching a code snippet related to this:

```
def __init__(self,input_channels,output_channels):
          self.conv1=GCNConv(output_channels,output_channels,node_dim=1,bias=False,normalize=False)
          self.conv1.weight=nn.Parameter(torch.ones_like(self.conv1.weight,dtype=torch.float64))
          self.register_parameter('edge_weight',None)

def reset_parameters(self,edge_index_t):
	 self.edge_weight = nn.Parameter(torch.ones((edge_index_t.size(1),),dtype=torch.float64,requires_grad=True))
def forward(self,data):
      x,edge_index_t = data.x,data.edge_index
       edge_index_t,__ =utils.add_self_loops(edge_index_t)
      if(self.edge_weight is None):
	   self.reset_parameters(edge_index_t)
      x=x.permute(1,0,2).contiguous()
      x=self.conv1(x,edge_index_t,edge_weight=self.edge_weight)
      x=x.permute(1,0,2).contiguous()
     data.x = x
    return data
```
As far as I can understand, the edge_index parameter is used to generate the adjacency matrix that determines which node's features we are supposed to consider for convolution. In the pytorch geometric documentation, [https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.GCNConv](url) , it is mentioned that the values of the adjacency matrix can be set to values other than 1 by the edge_weight. **My intention is to make the values present in the adjacency matrix learnable.** When I tried executing the loss.backward() and printed the values of self.edge_weight, the gradients were computed for it but the weights weren't updated. 
How do I solve this?

**Comment by rusty1s** [comment]
That should be well possible. Did you include the edge weights as parameters to `Adam`?

**Comment by hariharannatesh** [comment]
I am using SGD optimizer. In that I am mentioning it as 
`optimizer=optim.SGD(model.parameters(),lr=0.1,momentum=0.9,weight_decay=0.0001)`
I could see the self.edge_weight parameter being printed when I run the epoch. But after every epoch, the self.edge_weight is always filled with 1's because I am initializing it to that in the self.reset_parameters(). This function gets called only once but the value is fixed and doesn't get updated by the gradients.

**Comment by rusty1s** [comment]
I cannot really reproduce this. This works just fine for me:
```
import torch
import torch.nn.functional as F
from torch_geometric.datasets import Planetoid
import torch_geometric.transforms as T
from torch_geometric.nn import GCNConv

dataset = Planetoid('/tmp/Planetoid', 'Cora', transform=T.NormalizeFeatures())
data = dataset[0]


class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.edge_weight = torch.nn.Parameter(torch.ones(data.num_edges))
        self.conv1 = GCNConv(dataset.num_features, 16)
        self.conv2 = GCNConv(16, dataset.num_classes)

    def forward(self, x, edge_index):
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.conv1(x, edge_index, self.edge_weight.sigmoid()).relu()
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.conv2(x, edge_index, self.edge_weight.sigmoid())
        return F.log_softmax(x, dim=1)


device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model, data = Net().to(device), data.to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.01, weight_decay=5e-4)


def train():
    model.train()
    optimizer.zero_grad()
    out = model(data.x, data.edge_index)
    F.nll_loss(out[data.train_mask], data.y[data.train_mask]).backward()
    optimizer.step()


@torch.no_grad()
def test():
    model.eval()
    logits, accs = model(data.x, data.edge_index), []
    for _, mask in data('train_mask', 'val_mask', 'test_mask'):
        pred = logits[mask].max(1)[1]
        acc = pred.eq(data.y[mask]).sum().item() / mask.sum().item()
        accs.append(acc)
    return accs


best_val_acc = test_acc = 0
for epoch in range(1, 201):
    train()
    train_acc, val_acc, tmp_test_acc = test()
    print(model.edge_weight[:5])
    if val_acc > best_val_acc:
        best_val_acc = val_acc
        test_acc = tmp_test_acc
    log = 'Epoch: {:03d}, Train: {:.4f}, Val: {:.4f}, Test: {:.4f}'
    print(log.format(epoch, train_acc, best_val_acc, test_acc))
```

**Comment by hariharannatesh** [comment]
Thank you for your response. 
The thing is if we are batching graphs with a variable number of nodes and pass those batches for training, then it would not be feasible to declare the self.edge_weight parameter within the Net() class because it would be dependent on the edge_index that itself would be variable. That is why I opted for the self.register_parameter() method so that based on the number of edges, the parameter could be initialized. 
Is it possible to do this? i.e making the edge_weight parameter learnable for a batch of graphs with a variable number of nodes?

**Comment by rusty1s** [comment]
I see, this is indeed tricky. You cannot simply treat trainable parameters as `data` in PyTorch. An alternative is to create a huge Parameter tensor (for every edge that exists in the dataset), and to then select the edge weights you need during training:
```python
edge_weight = Parameter(torch.ones(num_all_edges))

def forward(self, x, edge_index, e_id):
    current_edge_weight = edge_weight[e_id]
    ...
```
where `e_id` denotes the ids of edges for that single forward pass.

**Comment by Volodimirich** [comment]
`
>         x = self.conv1(x, edge_index, self.edge_weight.sigmoid()).relu()
`
A similar problem occurred, this solution helped, but a small question remained. Why do we need sigmoid on edge_weight?

**Comment by rusty1s** [comment]
`GCNConv` can only handle positive edge weights, so you would need to have some pre-processing on `edge_weight` to guarantee that - `sigmoid` is one approach with the nice benefit that edge weights will be normalized to `[0, 1]`.

**Comment by Volodimirich** [comment]
Get it, thank you!)

**Comment by Wienhannover** [comment]
Hi, when it comes to interpret a trained GNN model, is there any difference between using learnable edge weights, such as 
#------------------------------------------------------------------------------
masked_prediction = model_to_explain(feats, graph, edge_weights=self.edge_mask) 
#------------------------------------------------------------------------------
and 
#------------------------------------------------------------------------------
self.edge_mask = torch.nn.Parameter(torch.randn(E) * std)
for module in self.model_to_explain.modules():
    if isinstance(module, MessagePassing):
        module.__explain__ = True
        module.__edge_mask__ = self.edge_mask
masked_prediction = model_to_explain(feats, graph)
#------------------------------------------------------------------------------
?

**Comment by zhenzhang1998** [comment]
Can a separate edge weight be set for each type of edge? Currently, the code above sets a weight parameter for each individual edge. Is it possible for edges of the same type to have the same weight?

**Comment by JiChern** [comment]
> `GCNConv` can only handle positive edge weights, so you would need to have some pre-processing on `edge_weight` to guarantee that - `sigmoid` is one approach with the nice benefit that edge weights will be normalized to `[0, 1]`.

So is there any possible way to learn a graph that both contains positive and negative edge weights?
